### PR TITLE
Added mx.stack c++ frontend impl

### PR DESCRIPTION
**Pull request by j-csc**
This pr will add `stack` support for `mlx.core` c++ frontend.

```cpp
// Functions:
array stack(const std::vector<array>& arrays, int axis, StreamOrDevice s = {});
array stack(const std::vector<array>& arrays, StreamOrDevice s = {});

// Usage:
auto x = array({1, 2, 3}, {3});
auto y = array({4, 5, 6}, {3});
auto z = std::vector<array>{x, y};

stack(z, axis=1) // shape (3,2)
```

Will add Python bindings in next PR. Also might need to work on making it a `Primitive` if needed @awni

### CODE REVIEW

File: mlx/utils.cpp
Comment (angeloskath) [inline mlx/utils.cpp:57]: I would change the name to `is_same_shape` and remove the dtype check which isn't in the name either.

File: mlx/ops.cpp
Comment (j-csc) [reply mlx/ops.cpp:609]: made the change in concat, used `result_type`

**Review by angeloskath — CHANGES_REQUESTED** [review]
Thanks for the changes. It looks great. If it isn't too much trouble I will ask for a few more updates. I hope it 's ok.

- Could you rebase on top of main and run the style checks?
- Could you also implement the python bindings? It is quite a simple addition after the work you 've already done.

Thanks again!

**Comment by j-csc** [comment]
pushed up the fix, lmk if there's anything else

File: python/src/ops.cpp
Comment (j-csc) [reply python/src/ops.cpp:2171]: will get these fixed up

File: python/src/ops.cpp
Comment (angeloskath) [inline python/src/ops.cpp]: I am really sorry to have to do this to you again but could you fix the style here. Basically `input` should be indented and stream should be in the line below.

File: python/src/ops.cpp
Comment (angeloskath) [inline python/src/ops.cpp]: Same as below style change, could you put it all in one line to match the rest of the signatures? Maybe in the future we 'll enforce all of them in 80 or 120 chars but for now...

File: python/src/ops.cpp
Comment (angeloskath) [inline python/src/ops.cpp]: Indent or even better simplify to `array: The resulting stacked array`

**Review by angeloskath — CHANGES_REQUESTED** [review]
Looks perfect. I left some style comments that currently break the docs if not updated.

Sorry for the back and forth. In the future if you want I could apply these minor changes myself and then merge. Just let me know if it is ok. I didn't do it by default cause I would show up as a co-author and steal a bit of your thunder which I didn't want.

**Comment by j-csc** [comment]
> Looks perfect. I left some style comments that currently break the docs if not updated.
> 
> Sorry for the back and forth. In the future if you want I could apply these minor changes myself and then merge. Just let me know if it is ok. I didn't do it by default cause I would show up as a co-author and steal a bit of your thunder which I didn't want.

Not at all, just happy to help! Will push up these fixes

**Comment by angeloskath** [comment]
@awni waiting on your approval on this.

**Review by awni — APPROVED** [review]

**Pull request merged**
